Refresh prompt after cd into folders

rumia-san opened the pull request:
Hi Vifon,
It is a great plugin for zsh. I think it is very useful and convenient.
But I found that the **prompt does not refresh** after cd into folders via "c" or "C" key.
Here is the snapshot for:
![old_behavior](https://user-images.githubusercontent.com/51704722/72162832-9d7a4580-33fd-11ea-837a-5a071c067f2c.png)
After "chdir" using "c" ro "C" key, the result of "pwd" command shows that the working directory has already changed. But the **prompt does not change**, which is confusing.
So I refresh the prompt via command
`zle --reset-prompt`
Then here is the screenshot for new behavior:
![new_behavior](https://user-images.githubusercontent.com/51704722/72163126-3315d500-33fe-11ea-83e5-1101cc2a457e.png)

Reference: [zsh: refresh prompt after running zle widget](https://stackoverflow.com/questions/52325626/zsh-refresh-prompt-after-running-zle-widget)

Vifon commented:
That's quite strange, as [zle redisplay](https://github.com/Vifon/deer/blob/a2eebadd3ccd64d9f269a49b3555b94b5aa5eed4/deer#L238) should already handle it. May I see your zsh prompt code?

If that's an actual bug, it'd be the best to move your `zsh reset-prompt` (I believe `--` is unnecessary here) into the `deer-restore` function to make sure it is always called.

rumia-san commented:
> That's quite strange, as [zle redisplay](https://github.com/Vifon/deer/blob/a2eebadd3ccd64d9f269a49b3555b94b5aa5eed4/deer#L238) should already handle it. May I see your zsh prompt code?
> 
> If that's an actual bug, it'd be the best to move your `zsh reset-prompt` (I believe `--` is unnecessary here) into the `deer-restore` function to make sure it is always called.

Hi Vifon,
I have confirmed that this bug was caused by another plugin **prezto**.

I removed my .zshrc, and load deer via antigen, The prompt worked well.
**After I loaded the prezto with lots of settings, this bug was reproduced.**
Here are the commands.
```
source ./antigen/antigen.zsh
zstyle ':prezto:*:*' color 'yes'
zstyle ':prezto:module:editor' key-bindings 'emacs'
zstyle ':prezto:module:git:alias' skip 'yes'
zstyle ':prezto:module:prompt' theme 'redhat'
zstyle ':prezto:module:prompt' pwd-length 'short'
zstyle ':prezto:module:terminal' auto-title 'yes'
zstyle ':prezto:module:autosuggestions' color 'yes'
zstyle ':prezto:module:python' autovenv 'yes'
zstyle ':prezto:load' pmodule \
	'environment' \
	'editor' \
	'history' \
	'git' \
	'utility' \
	'completion' \
	'history-substring-search' \
	'autosuggestions' \
	'prompt' \

	# 'autosuggestions' \

# Initialize prezto
antigen use prezto
antigen use Vifon/deer
autoload -U deer
zle -N deer
bindkey '\ek' deer
```

I have moved the "zle reset-prompt" to deer-restore() and remove the unnecessary "--" as well. Thanks for reminding : )

Vifon commented:
> I have confirmed that this bug was caused by another plugin prezto.  
I removed my .zshrc, export $PS1 manually, and load deer via antigen, The prompt worked well.
After I loaded the prezto with lots of settings, this bug was reproduced.
Here are the commands.

If that's the case, I believe it should be fixed in prezto, not in deer, as more plugins may be affected. Did you report it to them too? Doing it in deer would be merely a workaround.

rumia-san commented:
> If that's the case, I believe it should be fixed in prezto, not in deer, as more plugins may be affected. Did you report it to them too? Doing it in deer would be merely a workaround.

I have done a further investigation. I removed the prezto settings one by one and check the results. 

Finally, I found that this problem **is actually caused by [zsh-autosuggestions](https://github.com/zsh-users/zsh-autosuggestions)**, which is loaded by prezto

The **prezto itself works well with deer**, but in my settings use it to load zsh-autosuggestions plugin, which causes this problem.

**I have tried load the zsh-autosuggestions manually (source zsh-autosuggestions.zsh manually
) without prezto, and this problem occurs.**

It seems that the **zsh-autosuggestions will use/modify the edit buffer**.
The `zle redisplay` will **only redisplay** the edit buffer, while `zle reset-prompt` will **force the prompt to be re-expanded** and then redisplay the edit buffer. I guess that is the root cause. But I am new to zsh and I am unable to understand the workflow of zsh widgets/edit buffers/ zsh scripts.

> redisplay (unbound) (^R) (^R)
> Redisplays the edit buffer.
> 
> reset-prompt (unbound) (unbound) (unbound)
> Force the prompts on both the left and right of the screen to be re-expanded, then redisplay the edit buffer. This reflects changes both to the prompt variables themselves and changes in the expansion of the values (for example, changes in time or directory, or changes to the value of variables referred to by the prompt).
> 
> Otherwise, the prompt is only expanded each time zle starts, and when the display as been interrupted by output from another part of the shell (such as a job notification) which causes the command line to be reprinted.

Vifon commented:
The documentation seems to suggest `reset-prompt` is indeed more appropriate here and `redisplay` was only working by chance. Thanks for bringing it to my attention!

Please contact the zsh-autosuggestions team too, they may want to know the behavior change their plugin causes as I am not sure whether it is a bug or not.
